File: VB/HowToUseTreeListEditor.Module/CategoryWithIssues.vb
Imports DevExpress.Xpo
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Base.General
Imports System.ComponentModel

Namespace HowToUseTreeListEditor.Module

    <NavigationItem>
    Public MustInherit Class CategoryWithIssues
        Inherits BaseObject
        Implements ITreeNode

        <Association("CategoryWithIssues-Issues")>
        Public ReadOnly Property Issues As XPCollection(Of Issue)
            Get
                Return GetCollection(Of Issue)("Issues")
            End Get
        End Property

        Private allIssuesField As XPCollection(Of Issue)

        Public ReadOnly Property AllIssues As XPCollection(Of Issue)
            Get
                If allIssuesField Is Nothing Then
                    allIssuesField = New XPCollection(Of Issue)(Session, False)
                    CollectIssuesRecursive(Me, allIssuesField)
                    allIssuesField.BindingBehavior = CollectionBindingBehavior.AllowNone
                End If

                Return allIssuesField
            End Get
        End Property

        Private Sub CollectIssuesRecursive(ByVal issueCategory As CategoryWithIssues, ByVal target As XPCollection(Of Issue))
            target.AddRange(issueCategory.Issues)
            For Each childCategory As CategoryWithIssues In issueCategory.ChildrenProp
                CollectIssuesRecursive(childCategory, target)
            Next
        End Sub

        Private nameField As String

        Protected MustOverride ReadOnly Property ParentProp As ITreeNode

        Protected MustOverride ReadOnly Property ChildrenProp As IBindingList

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Property NameProp As String
            Get
                Return nameField
            End Get

            Set(ByVal value As String)
                SetPropertyValue("Name", nameField, value)
            End Set
        End Property

#Region "ITreeNode"
        Private ReadOnly Property Children As IBindingList Implements ITreeNode.Children
            Get
                Return ChildrenProp
            End Get
        End Property

        Private ReadOnly Property Name As String Implements ITreeNode.Name
            Get
                Return NameProp
            End Get
        End Property

        Private ReadOnly Property Parent As ITreeNode Implements ITreeNode.Parent
            Get
                Return ParentProp
            End Get
        End Property
#End Region
    End Class

    Public Class ProjectGroupWithIssues
        Inherits CategoryWithIssues

        Protected Overrides ReadOnly Property ParentProp As ITreeNode
            Get
                Return Nothing
            End Get
        End Property

        Protected Overrides ReadOnly Property ChildrenProp As IBindingList
            Get
                Return ProjectsWithIssues
            End Get
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Sub New(ByVal session As Session, ByVal name As String)
            MyBase.New(session)
            NameProp = name
        End Sub

        <Association("ProjectGroupWithIssues-ProjectsWithIssues"), Aggregated>
        Public ReadOnly Property ProjectsWithIssues As XPCollection(Of ProjectWithIssues)
            Get
                Return GetCollection(Of ProjectWithIssues)("ProjectsWithIssues")
            End Get
        End Property
    End Class

    Public Class ProjectWithIssues
        Inherits CategoryWithIssues

        Private projectGroupWithIssuesField As ProjectGroupWithIssues

        Protected Overrides ReadOnly Property ParentProp As ITreeNode
            Get
                Return projectGroupWithIssuesField
            End Get
        End Property

        Protected Overrides ReadOnly Property ChildrenProp As IBindingList
            Get
                Return ProjectAreasWithIssues
            End Get
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Sub New(ByVal session As Session, ByVal name As String)
            MyBase.New(session)
            NameProp = name
        End Sub

        <Association("ProjectGroupWithIssues-ProjectsWithIssues")>
        Public Property ProjectGroupWithIssues As ProjectGroupWithIssues
            Get
                Return projectGroupWithIssuesField
            End Get

            Set(ByVal value As ProjectGroupWithIssues)
                SetPropertyValue("ProjectGroupWithIssues", projectGroupWithIssuesField, value)
            End Set
        End Property

        <Association("ProjectWithIssues-ProjectAreasWithIssues"), Aggregated>
        Public ReadOnly Property ProjectAreasWithIssues As XPCollection(Of ProjectAreaWithIssues)
            Get
                Return GetCollection(Of ProjectAreaWithIssues)("ProjectAreasWithIssues")
            End Get
        End Property
    End Class

    Public Class ProjectAreaWithIssues
        Inherits CategoryWithIssues

        Private projectWithIssuesField As ProjectWithIssues

        Protected Overrides ReadOnly Property ParentProp As ITreeNode
            Get
                Return projectWithIssuesField
            End Get
        End Property

        Protected Overrides ReadOnly Property ChildrenProp As IBindingList
            Get
                Return New BindingList(Of Object)()
            End Get
        End Property

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Public Sub New(ByVal session As Session, ByVal name As String)
            MyBase.New(session)
            NameProp = name
        End Sub

        <Association("ProjectWithIssues-ProjectAreasWithIssues")>
        Public Property ProjectWithIssues As ProjectWithIssues
            Get
                Return projectWithIssuesField
            End Get

            Set(ByVal value As ProjectWithIssues)
                SetPropertyValue("ProjectWithIssues", projectWithIssuesField, value)
            End Set
        End Property
    End Class
End Namespace

File: CS/HowToUseTreeListEditor.Module/CategoryWithIssues.cs
using System;

using DevExpress.Xpo;

using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;
using DevExpress.Persistent.Base.General;
using System.ComponentModel;

namespace HowToUseTreeListEditor.Module {
    [NavigationItem]
    public abstract class CategoryWithIssues : BaseObject, ITreeNode {
        [Association("CategoryWithIssues-Issues")]
        public XPCollection<Issue> Issues {
            get {
                return GetCollection<Issue>("Issues");
            }
        }
        private XPCollection<Issue> allIssues;
        public XPCollection<Issue> AllIssues {
            get {
                if(allIssues == null) {
                    allIssues = new XPCollection<Issue>(Session, false);
                    CollectIssuesRecursive(this, allIssues);
                    allIssues.BindingBehavior = CollectionBindingBehavior.AllowNone;
                }
                return allIssues;
            }
        }
        private void CollectIssuesRecursive(CategoryWithIssues issueCategory, XPCollection<Issue> target) {
            target.AddRange(issueCategory.Issues);
            foreach(CategoryWithIssues childCategory in issueCategory.Children) {
                CollectIssuesRecursive(childCategory, target);
            }
        }
        private string name;
        protected abstract ITreeNode Parent {
            get;
        }
        protected abstract IBindingList Children {
            get;
        }
        public CategoryWithIssues(Session session) : base(session) { }
        public string Name {
            get {
                return name;
            }
            set {
                SetPropertyValue("Name", ref name, value);
            }
        }
        #region ITreeNode
        IBindingList ITreeNode.Children {
            get {
                return Children;
            }
        }
        string ITreeNode.Name {
            get {
                return Name;
            }
        }
        ITreeNode ITreeNode.Parent {
            get {
                return Parent;
            }
        }
        #endregion
    }
    public class ProjectGroupWithIssues : CategoryWithIssues {
        protected override ITreeNode Parent {
            get {
                return null;
            }
        }
        protected override IBindingList Children {
            get {
                return ProjectsWithIssues;
            }
        }
        public ProjectGroupWithIssues(Session session) : base(session) { }
        public ProjectGroupWithIssues(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("ProjectGroupWithIssues-ProjectsWithIssues"), Aggregated]
        public XPCollection<ProjectWithIssues> ProjectsWithIssues {
            get {
                return GetCollection<ProjectWithIssues>("ProjectsWithIssues");
            }
        }
    }

    public class ProjectWithIssues : CategoryWithIssues {
        private ProjectGroupWithIssues projectGroupWithIssues;
        protected override ITreeNode Parent {
            get {
                return projectGroupWithIssues;
            }
        }
        protected override IBindingList Children {
            get {
                return ProjectAreasWithIssues;
            }
        }
        public ProjectWithIssues(Session session) : base(session) { }
        public ProjectWithIssues(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("ProjectGroupWithIssues-ProjectsWithIssues")]
        public ProjectGroupWithIssues ProjectGroupWithIssues {
            get {
                return projectGroupWithIssues;
            }
            set {
                SetPropertyValue("ProjectGroupWithIssues", ref projectGroupWithIssues, value);
            }
        }
        [Association("ProjectWithIssues-ProjectAreasWithIssues"), Aggregated]
        public XPCollection<ProjectAreaWithIssues> ProjectAreasWithIssues {
            get {
                return GetCollection<ProjectAreaWithIssues>("ProjectAreasWithIssues");
            }
        }
    }

    public class ProjectAreaWithIssues : CategoryWithIssues {
        private ProjectWithIssues projectWithIssues;
        protected override ITreeNode Parent {
            get {
                return projectWithIssues;
            }
        }
        protected override IBindingList Children {
            get {
                return new BindingList<object>();
            }
        }
        public ProjectAreaWithIssues(Session session) : base(session) { }
        public ProjectAreaWithIssues(Session session, string name)
            : base(session) {
            this.Name = name;
        }
        [Association("ProjectWithIssues-ProjectAreasWithIssues")]
        public ProjectWithIssues ProjectWithIssues {
            get {
                return projectWithIssues;
            }
            set {
                SetPropertyValue("ProjectWithIssues", ref projectWithIssues, value);
            }
        }
    }
}
